### PR TITLE
Create new Docker publishing workflow.

### PR DESCRIPTION
**Pull request by cgranade**
This PR sets up a new GitHub Action to prebuild the Docker image for the samples repo and publish it to the packages feed for this repo.

### CODE REVIEW

**Comment by cgranade** [comment]
So the infrastructure for this works out to be pretty straightforward, and seems to work pretty well at getting Docker images from master up. The trouble is that apparently GitHub Package Registry has an undocumented requirement that users log in to `docker pull` even public packages from public repos (more details at https://github.community/t5/GitHub-Actions/docker-pull-from-public-GitHub-Package-Registry-fail-with-quot/td-p/32782). This means that the new images deploy correctly, but are difficult or impossible to use with remote development, VS Online, or Binder. Since those are our three biggest usecases, I will work to explore other solutions.

+@anpaz-msft

File: Dockerfile [hunk 1]
```diff
@@ -1,46 +1,10 @@
...
-# https://github.com/microsoft/iqsharp/blob/master/images/iqsharp-base/Dockerfile.
-# As per Binder documentation, we choose to use an SHA sum here instead of a
-# tag.
-FROM mcr.microsoft.com/quantum/iqsharp-base:0.10.1912.501
+# This uses the latest Docker image built from the samples repository,
+# defined by the Dockerfile in Build/images/samples.
+FROM docker.pkg.github.com/microsoft/quantum/samples:latest
 
 # Mark that this Dockerfile is used with the samples repository.
 ENV IQSHARP_HOSTING_ENV=SAMPLES_DOCKERFILE
```
Comment (anpaz-msft) [inline Dockerfile]: I wonder if we should have different values for both, maybe this one should be SAMPLES_HOSTED?

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    - name: Azure login
+      uses: Azure/login@v1
+      with:
+        creds: ${{ secrets.ACR_CREDENTIALS }}
+    - name: Docker Login
+      run: az acr login -n msquantum
+    - name: Build and push Docker image
+      run: |
+        $Now = [DateTime]::Now;
+        $ImageTag = "{0}{1:d2}{2:d2}.{3:d2}{4:d2}.{5:d2}{6:d4}" -f $Now.Year, $Now.Month, $Now.Day, $Now.Hour, $Now.Minute, $Now.Second, $Now.Millisecond
```
Comment (anpaz-msft) [inline .github/workflows/main.yml]: would it be better to use the commit id?

File: Dockerfile (hunk 1 above)
Comment (cgranade) [reply Dockerfile]: Good point, easy enough to change, yeah.

File: .github/workflows/main.yml (hunk 2 above)
Comment (cgranade) [reply .github/workflows/main.yml]: Happy to do that as well, it's easily accessible as `${{ github.sha }}`. My only thought was that this gives a strictly increasing ordering. That said, we likely won't need to directly refer to tags other than latest, so let me switch over to SHA sums.

**Pull request merged**